src/data/matrix/kronecker.lean

```diff
@@ -153,6 +153,18 @@ begin
   simp_rw [f.map_sum, linear_map.sum_apply, linear_map.map_sum, h_comm],
 end
 
+lemma kronecker_map_assoc {δ ξ ω ω' : Type*} (f : α → β → γ) (g : γ → δ → ω) (f' : α → ξ → ω')
+  (g' : β → δ → ξ) {q r : Type*} [fintype q] [fintype r] (A : matrix l m α) (B : matrix n p β)
+  (D : matrix q r δ) (φ : ω ≃ ω') (hφ : ∀ a b d, φ (g (f a b) d ) = f' a (g' b d )) :
+  (reindex (equiv.prod_assoc l n q) (equiv.prod_assoc m p r)).trans (equiv.map_matrix φ)
+    (kronecker_map g (kronecker_map f A B) D) = kronecker_map f' A (kronecker_map g' B D) :=
```

Comment:
I'm not convinced `φ` is helpful here, as it can just be folded into `g`; why not:
```suggestion
lemma kronecker_map_assoc {δ ξ ω : Type*} (f : α → β → γ) (g : γ → δ → ω) (f' : α → ξ → ω)
  (g' : β → δ → ξ) {q r : Type*} [fintype q] [fintype r] (A : matrix l m α) (B : matrix n p β)
  (D : matrix q r δ) (h : ∀ a b d, g (f a b) d = f' a (g' b d)) :
  (reindex (equiv.prod_assoc l n q) (equiv.prod_assoc m p r)).trans (equiv.map_matrix φ)
    (kronecker_map g (kronecker_map f A B) D) = kronecker_map f' A (kronecker_map g' B D) :=
```

Reply:
Well, because from a mathematical point of view I would see `g` and `φ` as different objects: one is an equivalence between the landing types, the other is really part of the data. I agree that one can pack one into the other, but I don't find it natural.

Reply:
I just edited the above - the argument for removing `φ` is that it makes the statement simpler, as the `equiv.map_matrix` is no longer on the LHS. We already have lemmas for how `kronecker_map` and `map` interact; it's usually better to have lemmas for simple statements then combine them to form the complex statements, rather than having complex lemmas for which the simple lemmas fall out as special cases: consider how annoying having only `mul_comm_comm : a * (b * c) = (c * b) * a` would be, and being told "oh it's easy, just set `a` to `1` to get the special case!"

Reply:
From a resolved comment below:

> As far as `φ` is concerned, I am not sure to understand what you mean but as said I somewhat prefer the version with it as it seems more natural to me.

My point is that if you faced with a choice between a lemma with a long statement and a lemma with a short statement, you should usually prove the short one first and then use the short one to prove the long one.

In this case, the short one is the version without `φ`, the long one is the current version, and the way to prove one from the other is `rw [kronecker_map_map, the_short_version]`.

Reply:
Uhm... this sounds strange to me: I would rather usually prove a more general statement and, if needed for some practical purposes, would deduce the short one from the general one. Moreover, I can't make your version work (`dsimp; rw [kronecker_map_map, kronecker_map_assoc_eric].` complains about `rw`); also, this proof is more verbose than `ext $ λ i j, hφ _ _ _` which also works very well for the general version, so the short one is not really useful to prove the general one. If you want we can add the short version, but I find it a bit useless.

Reply:
If you don't want to remove `φ`, it might make sense to relax it to be any function:
```suggestion
lemma kronecker_map_assoc {δ ξ ω ω' : Type*} (f : α → β → γ) (g : γ → δ → ω) (f' : α → ξ → ω')
  (g' : β → δ → ξ) (A : matrix l m α) (B : matrix n p β) (D : matrix q r δ) (φ : ω → ω')
  (hφ : ∀ a b d, φ (g (f a b) d ) = f' a (g' b d )) :
  reindex (equiv.prod_assoc l n q) (equiv.prod_assoc m p r)
    ((kronecker_map g (kronecker_map f A B) D).map φ) = kronecker_map f' A (kronecker_map g' B D) :=
```
Another option might just be to remove this proof in its entirety, as the name is longer than the proof.

I'll leave it to @jcommelin or @Vierkantor to make the final call.

Reply:
If the general statement has an easier proof, then why not keep it, and then add a specialization where `φ := id`, which will look something like @eric-wieser's original suggestion? I don't think adding more lemmas is going to cause any pain in the long run.

Reply:
The proofs for the longer and shorter statements are the same.

Reply:
OK, I have added `lemma kronecker_map_assoc₁` for the special case where `φ := id`. I also have another remark: all lemmas marked with `foo_linear` actually concerns _bilinear_ maps and I'd like to rename them `bilinear`. I have already done this for `kronecker_bilinear` but there are still some for `kronecker_map`. Any objection?

Reply:
Changing to `_bilinear` is fine by me, I think I just missed those.

Reply:
Ok, the commit is coming.